core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -499,9 +500,25 @@ private synchronized void ensureSnapshotsLoaded() {
       List<Snapshot> loadedSnapshots = Lists.newArrayList(snapshotsSupplier.get());
       loadedSnapshots.removeIf(s -> s.sequenceNumber() > lastSequenceNumber);
 
-      // Format version 1 does not have accurate sequence numbering, so remove based on timestamp
+      // Format version 1 does not have accurate sequence numbering, so remove based on
+      // ref reachability
       if (this.formatVersion == 1) {
-        loadedSnapshots.removeIf(s -> s.timestampMillis() > currentSnapshot().timestampMillis());
+        Set<Long> snapshotsToKeep = Sets.newHashSet();
+
+        for (Long snapshotId :
+            refs().values().stream().map(SnapshotRef::snapshotId).collect(Collectors.toSet())) {
+          SnapshotUtil.ancestorsOf(
```

Comment:
worth mentioning that this will include the starting `snapshotId`

Reply:
Also got bit by that one lol: https://github.com/apache/iceberg/issues/6930